awx/api/views.py

```diff
@@ -2748,17 +2748,14 @@ def post(self, request, *args, **kwargs):
         prompted_fields = _accepted_or_ignored[0]
```

Comment:
See the line above

```python
_accepted_or_ignored = obj._accept_or_ignore_job_kwargs(**request.data)
```

Note that this is using `request.data`. It is not using `serializer.data`. When you do the translation of credential / vault_credential / etc. into `credentials`, you do it inside of the `to_internal_value` inside of the serializer. But in order for that to work, the output for `to_internal_value` needs to become the ultimate source of truth for the translated request data. If `request.data` is passed around after the translation is done on some _other_ dictionary, then it's going to cause bugs.

Reply:
👍 what do you recommend here?  It made sense to me to do this at the serializer level because at that point we have access to the underlying JT, which we need to compose the translated `.credentials`.  Doing this at the `request.data` level in a view is probably going to be quite a bit trickier.

Reply:
I guess you _could_ argue for loading the object and doing this in the views.  It's how we handled similar work for `extra_credentials`:

https://github.com/ansible/tower/blob/devel/awx/api/views.py#L2730

Reply:
The line you linked to looks like it was handling the vault_credenial_id -> vault_credential conversion. We even hacked the mutability of request.data... oof.

I would recommend that either option is valid:
 - moving the new translations (stuffing stuff into `credentials`) to the view, modifying `request.data`
 - moving the existing translations into `to_internal_value` and replace `request.data` with `serializer.data`

If you did the latter option, you would be navigating into uncharted waters. I don't know what unexpected ways in which the serializer might modify the data... but it's the better option in the big picture. So it's up to you.

Reply:
Okay, addressed in a new patch 👍